File: backend/src/main/java/com/stockanalysis/backtest/spec/StrategySpec.java
package com.stockanalysis.backtest.spec;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;

/**
 * Full strategy rule spec. This is exactly the JSON persisted in {@code strategy.spec_json}
 * and the format the (future) LLM recommender produces.
 */
@JsonIgnoreProperties(ignoreUnknown = true)
public class StrategySpec {

    private String name;
    private String source = "USER";
    private Position position = Position.LONG;
    private TargetType targetType = TargetType.SINGLE;
    private ConditionGroup entry = new ConditionGroup();
    private ExitSpec exit = new ExitSpec();
    private PremarketSpec premarket = new PremarketSpec();
    private CapitalSpec capital = new CapitalSpec();

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getSource() {
        return source;
    }

    public void setSource(String source) {
        this.source = source;
    }

    public Position getPosition() {
        return position;
    }

    public void setPosition(Position position) {
        this.position = position == null ? Position.LONG : position;
    }

    public ConditionGroup getEntry() {
        return entry;
    }

    public void setEntry(ConditionGroup entry) {
        this.entry = entry;
    }

    public ExitSpec getExit() {
        return exit;
    }

    public void setExit(ExitSpec exit) {
        this.exit = exit;
    }

    public TargetType getTargetType() {
        return targetType;
    }

    public void setTargetType(TargetType targetType) {
        this.targetType = targetType == null ? TargetType.SINGLE : targetType;
    }

    public PremarketSpec getPremarket() {
        return premarket;
    }

    public void setPremarket(PremarketSpec premarket) {
        this.premarket = premarket;
    }

    public CapitalSpec getCapital() {
        return capital;
    }

    public void setCapital(CapitalSpec capital) {
        this.capital = capital == null ? new CapitalSpec() : capital;
    }

    /** True when this strategy should run in the ETF pre-market futures-trend mode. */
    public boolean usesPremarket() {
        return targetType == TargetType.ETF && premarket != null && premarket.isEnabled();
    }
}

File: backend/src/main/kotlin/com/stockanalysis/backtest/spec/StrategySpec.kt
package com.stockanalysis.backtest.spec

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonSetter
import com.fasterxml.jackson.annotation.Nulls

/**
 * 매매 로직 전체. **이 프로젝트에서 규칙의 유일한 원본**입니다.
 *
 * `strategy.spec_json`에 그대로 저장되고, LLM 추천도 같은 포맷을 만들며, 무엇보다
 * **백테스트·모의투자·실전투자가 전부 이 하나를 읽습니다**. 실투자 쪽에 익절·손절 같은 필드를
 * 복제해두면 두 벌이 되는 순간 어긋나고, 그러면 백테스트가 의미를 잃습니다.
 *
 * 기본값이 있는 필드를 새로 추가할 때 주의: 예전에 저장된 행에는 그 키가 아예 없어서
 * Hibernate가 재직렬화 결과를 다르게 보고 **조용한 UPDATE**를 날립니다. 쓰기 트랜잭션에서
 * 전략을 읽는 곳은 반드시 `entityManager.detach`를 거쳐야 합니다(`BacktestService` 참고).
 */
@JsonIgnoreProperties(ignoreUnknown = true)
class StrategySpec {

    var name: String? = null

    /** USER 또는 LLM. 누가 만든 규칙인지 표시용입니다. */
    var source: String? = "USER"

    // 아래 필드들은 전부 `nulls = SKIP`입니다. 기본값이 있는 필드를 나중에 추가했기 때문에
    // 예전에 저장된 spec_json에는 키가 없거나 null인 행이 있고, Kotlin 비-널 타입은 null을
    // 거부하므로 Jackson 단계에서 흘려보내 기본값을 살립니다.

    @JsonSetter(nulls = Nulls.SKIP)
    var position: Position = Position.LONG

    @JsonSetter(nulls = Nulls.SKIP)
    var targetType: TargetType = TargetType.SINGLE

    @JsonSetter(nulls = Nulls.SKIP)
    var entry: ConditionGroup = ConditionGroup()

    @JsonSetter(nulls = Nulls.SKIP)
    var exit: ExitSpec = ExitSpec()

    @JsonSetter(nulls = Nulls.SKIP)
    var premarket: PremarketSpec = PremarketSpec()

    @JsonSetter(nulls = Nulls.SKIP)
    var capital: CapitalSpec = CapitalSpec()

    /**
     * ETF 장전추세 모드로 돌려야 하는 전략인지.
     *
     * 이 모드에서는 [entry] 조건이 **무시됩니다** — 그날 무엇을 살지는 장전 선물 추세가 정하고,
     * 매수 시점도 장 시작으로 고정됩니다.
     */
    fun usesPremarket(): Boolean = targetType == TargetType.ETF && premarket.enabled
}
